### PR TITLE
Write buffer to a temp file rather than silently forcing save

### PR DESCRIPTION
**Pull request by purcell**
This is an initial solution to #242, wherein `intero`'s magical silent saving of buffers was causing all sorts of problems.

The implementation here uses a buffer-local variable to track an associated temp file, the contents of which will be kept up-to-date with the buffer contents. Intero backend calls are then made against that buffer, and the flycheck results translated back accordingly.

With these changes, all save-related behaviour (including before/after save hooks) works as expected.

AFAICT this all works fine. The implementation is slightly inefficient, in that it will sometimes lead to the buffer being redundantly re-written to the temp file, but it's fast and straightforward to understand. If you'd prefer, I'd be happy to work on streamlining it.

### CODE REVIEW

File: elisp/intero.el [hunk 1]
```diff
@@ -490,10 +458,9 @@ CHECKER and BUFFER are added to each item parsed from STRING."
                        line column type
                        msg
                        :checker checker
-                       :buffer (when (string= (intero-buffer-file-name buffer)
-                                              file)
+                       :buffer (when (string= temp-file file)
                                  buffer)
-                       :filename file)
+                       :filename (intero-buffer-file-name buffer))
```
Comment (chrisdone) [inline elisp/intero.el]: We may want to also do a search-and-replace for the temp filename in error messages, which GHC sometimes references.

**Comment by chrisdone** [comment]
Interesting approach. I like it.

**Comment by purcell** [comment]
Thanks for merging. If any issues arise from this, just @ tag me in, and I'll be happy to help.

**Comment by chrisdone** [comment]
Just tried it out, here's an example of the above:

``` haskell
 Par...   68  28 error           Couldn't match expected type ‘[t0] -> P (a SrcSpanInfo)’
                with actual type ‘P (a SrcSpanInfo)’
    Relevant bindings include
      p :: P (a SrcSpanInfo)
        (bound at /var/folders/2s/6j1qqrgd57bg9ctkltrkn3z80000gn/T/intero52962niH.hs:68:22)
      normalParserNoFixity :: P (a SrcSpanInfo)
                              -> Maybe [Fixity] -> P (a SrcSpanInfo)
        (bound at /var/folders/2s/6j1qqrgd57bg9ctkltrkn3z80000gn/T/intero52962niH.hs:68:1)
    The function ‘p’ is applied to one argument,
    but its type ‘P (a SrcSpanInfo)’ has none
    In the expression: p []
    In an equation for ‘normalParserNoFixity’:
        normalParserNoFixity p _ = p []... (intero)

```

If we pop in a search-and-replace for that filename, it would mitigate the noise, I think.

**Comment by shanemikel** [comment]
@purcell, do you mind taking a look at #279?

**Pull request merged**